# Analysis Code: a census of hidden and discoverable microbial diversity beyond genome-centric approaches

This repository contains analysis code for the manuscript on "A census of hidden and discoverable microbial diversity beyond genome-centric approaches" by [Prasoodanan, Maistrenko, et al](https://www.biorxiv.org/content/10.1101/2025.06.26.661807v1).

## Abstract

Cataloguing Earth’s biodiversity remains one of biology’s most formidable challenges, and the greatest diversity is expected to reside among the smallest organisms: microbes. Yet the ongoing census of microbial life is hampered by disparate sampling of Earth’s habitats, challenges in isolating uncultivated organisms, limited resolution in taxonomic marker gene amplicons, and incomplete recovery of metagenome-assembled genomes (MAGs). Here, we quantified discoverable bacterial and archaeal diversity in a comprehensive, curated cross-habitat dataset of 92,187 metagenomes. Clustering 502M sequences of 130 marker genes, we detected 705k bacterial and 27k archaeal species-level clades, the vast majority of which was hidden among ‘unbinned’ contigs, beyond current genome-centric approaches. At deeper taxonomic levels, we estimate that 10 archaeal and 145 bacterial novel phyla and around 80k novel genera are discoverable in current data. We identified soils and aquatic environments as novel lineage recovery hotspots, yet predict that discovery will remain in full swing across habitats as more data accrues.

## Description




## Availability of underlying data

Metagenomic assemblies, Metagenome-Assembled Genomes (MAGs), gene calls and corresponding annotations are available via the SPIRE db [downloads page](spire.embl.de/downloads).

Inferred marker gene phylogenies with annotations, as well as pre-generated tree visualizations for archaeal markers are available via the EBI BioStudies repository the following accessions:

| Accession | Description |
| ---- | ---- |
| [S-BSST2111](https://www.ebi.ac.uk/biostudies/studies/S-BSST2111) | 28 pre-generated plots of archaeal marker gene phylogenies, as PDF |
| [S-BSST2112](https://www.ebi.ac.uk/biostudies/studies/S-BSST2112) | 53 bacterial marker gene phylogenies in `nwk` format |
| [S-BSST2113](https://www.ebi.ac.uk/biostudies/studies/S-BSST2113) | Tip-level annotations and metadata for 53 bacterial marker gene phylogenies |
| [S-BSST2116](https://www.ebi.ac.uk/biostudies/studies/S-BSST2116) | Tip-level annotations and metadata for 28 bacterial marker gene phylogenies |
| [S-BSST2117](https://www.ebi.ac.uk/biostudies/studies/S-BSST2117) | 28 archaeal marker gene phylogenies in `nwk` format |

Additional data, e.g. sample-level annotations, are available as supplementary material to the [preprint](https://www.biorxiv.org/content/10.1101/2025.06.26.661807v1.supplementary-material) on bioRxiv.

## Contributors

Lead authors:

Vishnu Prasoodanan P K, APC Microbiome & School of Medicine, University College Cork, Ireland. [ORCID: 0009-0002-1872-5790](https://orcid.org/0009-0002-1872-5790)

Oleksandr M Maistrenko, Department of Marine Microbiology & Biogeochemistry, Royal Netherlands Institute for Sea Research(NIOZ), The Netherlands. [ORCID: 0000-0003-1961-7548](https://orcid.org/0000-0003-1961-7548).


Correspondence:

Thomas S B Schmidt, APC Microbiome & School of Medicine, University College Cork, Ireland. [ORCID: 0000-0001-8587-4177](https://orcid.org/0000-0001-8587-4177).

sebastian _dot_ schmidt _at_ ucc _dot_ ie

